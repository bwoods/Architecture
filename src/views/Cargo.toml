[package]
name = "composable_views"
version = "0.0.0"
edition = "2021"

[lib]


[features]
default_ui = []


[dependencies]
dependencies = { workspace = true }
rustybuzz = { version = "0.12.1" }
lyon = { version = "1.0.1", default-features = false }
usvg = { version = "0.37.0", default-features = false }
svg = { version = "0.14.0", default-features = false }

[dependencies.ttf-parser]
version = "*" # whatever rustybuzz is usingâ€¦
features = ["std"] # enable its `String` support


[dev-dependencies]
insta = "1.34.0"
