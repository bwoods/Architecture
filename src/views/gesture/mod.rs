use std::cell::Cell;

use crate::dependencies::DependencyDefault;

mod recognizer;
pub use recognizer::*;

mod tap;

#[derive(Copy, Clone, Eq, PartialEq)]
pub struct Id(pub(crate) std::num::NonZeroU128);

#[doc(hidden)]
impl TryFrom<u128> for Id {
    type Error = std::num::TryFromIntError;

    fn try_from(value: u128) -> Result<Self, Self::Error> {
        Ok(Self(value.try_into()?))
    }
}

#[non_exhaustive] // must use `State::default()`
#[derive(Copy, Clone, Default, Eq, PartialEq)]
pub struct Values {
    pub active: Option<Id>,
    pub hover: Option<Id>,
    pub focus: Option<Id>,
}

impl DependencyDefault for Values {}

/// The user interface state carried between cycles by the application.
///
/// ```
/// # use composable::views::gesture::{Id, State};
/// let state = State::default();
///
/// let mut values = state.get();
/// # let id: Id = 1u128.try_into().unwrap();
/// values.active = Some(id);
///
/// // â€¦
///
/// state.set(values);
///
/// ```
pub type State = Cell<Values>;

#[test]
fn confirm_id_niche_optimization() {
    assert_eq!(std::mem::size_of::<Id>(), std::mem::size_of::<u128>());
}
