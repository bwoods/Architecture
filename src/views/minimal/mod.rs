//! A minimal, _but viable_, user interface layer.

mod font;
pub use font::{with_default_font, Inter};

#[derive(Copy, Clone)]
/// `Accessibility` defines a predefined scale for scalable content.
pub enum Accessibility {
    /// Use extra extra small sizes.  
    /// This is the default on desktop platforms.
    XXS,
    /// use extra small sizes.
    XS,
    /// Use small sizes.
    S,
    /// Use medium sizes.
    M,
    /// Use large sizes.  
    /// This is the default on mobile platforms.
    L,
    /// use extra large sizes.
    XL,
    /// Use extra extra large sizes.
    XXL,
    /// Use extra extra extra large sizes.
    XXXL,
}
