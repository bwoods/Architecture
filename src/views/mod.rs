use std::ops::Deref;

pub use lyon::math::{Box2D as Bounds, Point, Size, Transform};

pub use layout::Spacer;
pub use modifiers::fixed::{Fixed, FixedHeight, FixedWidth};
pub use modifiers::padding::Padding;
pub use output::{gpu, svg, Output};
pub use shapes::{Circle, ContinuousRoundedRectangle, Ellipse, Rectangle, RoundedRectangle};
pub use shapes::{Path, Shape};
#[doc(inline)]
pub use text::Text;

use crate::{From, TryInto};

/// Alias for `euclid::default::SideOffsets2D<f32>`
pub type Offsets = lyon::geom::euclid::default::SideOffsets2D<f32>;

mod output;
/// Text handling for `View` construction.
pub mod text;

pub mod gesture;
mod layout;
mod modifiers;

mod shapes;
#[cfg(feature = "default_ui")]
pub mod ui;

/// User interface element and modifiers to re-configure it.
pub trait View: Sized {
    /// The intrinsic size of the `View`
    fn size(&self) -> Size;
    /// User-interface [`Event`] handling of the `View`
    #[allow(unused_variables)]
    fn event(&self, event: Event, offset: Point, bounds: Bounds) {}
    /// How the `View` is drawn
    fn draw(&self, bounds: Bounds, onto: &mut impl Output);

    /// Add padding to all sides of the `View`
    fn padding(self, top: f32, right: f32, bottom: f32, left: f32) -> Padding<Self> {
        Padding {
            offsets: Offsets::new(top, right, bottom, left),
            view: self,
        }
    }

    /// Add padding to the top of the `View`
    fn padding_top(self, pad: f32) -> Padding<Self> {
        self.padding(pad, 0.0, 0.0, 0.0)
    }

    /// Add padding to the right side of the `View`
    fn padding_right(self, pad: f32) -> Padding<Self> {
        self.padding(0.0, pad, 0.0, 0.0)
    }

    /// Add padding to the bottom of the `View`
    fn padding_bottom(self, pad: f32) -> Padding<Self> {
        self.padding(0.0, 0.0, pad, 0.0)
    }

    /// Add padding to the left side of the `View`
    fn padding_left(self, pad: f32) -> Padding<Self> {
        self.padding(0.0, 0.0, 0.0, pad)
    }

    /// Add padding to the horizontal sides of the `View`
    fn padding_horizontal(self, pad: f32) -> Padding<Self> {
        self.padding(0.0, pad, 0.0, pad)
    }

    /// Add padding to the vertical sides of the `View`
    fn padding_vertical(self, pad: f32) -> Padding<Self> {
        self.padding(0.0, 0.0, pad, 0.0)
    }

    /// Add different padding to the horizontal and vertical sides of the `View`
    fn padding_both(self, horizontal: f32, vertical: f32) -> Padding<Self> {
        self.padding(vertical, horizontal, vertical, horizontal)
    }

    /// Add the same padding to all sides of the `View`
    fn padding_all(self, pad: f32) -> Padding<Self> {
        self.padding(pad, pad, pad, pad)
    }

    /// Set the size of the `View` to a fixed value.
    fn fixed(self, width: f32, height: f32) -> impl View {
        let size = self.size();
        if size.is_empty() {
            return Err(Fixed {
                size: Size::new(width, height),
                view: self,
            });
        }

        let horizontal = (width - size.width) / 2.0;
        let vertical = (height - size.height) / 2.0;
        Ok(self.padding_both(horizontal, vertical))
    }

    fn width(self, width: f32) -> impl View {
        let size = self.size();
        if size.is_empty() {
            return Err(FixedWidth { width, view: self });
        }

        let horizontal = (width - size.width) / 2.0;
        Ok(self.padding_horizontal(horizontal))
    }

    fn height(self, height: f32) -> impl View {
        let size = self.size();
        if size.is_empty() {
            return Err(FixedHeight { height, view: self });
        }

        let vertical = (height - size.height) / 2.0;
        Ok(self.padding_vertical(vertical))
    }
}

impl<T: View> View for Box<T> {
    fn size(&self) -> Size {
        self.deref().size()
    }

    fn event(&self, event: Event, offset: Point, bounds: Bounds) {
        self.deref().event(event, offset, bounds)
    }

    fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
        self.deref().draw(bounds, onto)
    }
}

impl<T: View> View for Option<T> {
    fn size(&self) -> Size {
        if let Some(view) = self {
            return view.size();
        }

        Size::zero()
    }

    fn event(&self, event: Event, offset: Point, bounds: Bounds) {
        if let Some(view) = self {
            view.event(event, offset, bounds)
        }
    }

    fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
        if let Some(view) = self {
            view.draw(bounds, onto)
        }
    }
}

impl<T: View, E: View> View for Result<T, E> {
    fn size(&self) -> Size {
        match self {
            Ok(view) => view.size(),
            Err(view) => view.size(),
        }
    }

    fn event(&self, event: Event, offset: Point, bounds: Bounds) {
        match self {
            Ok(view) => view.event(event, offset, bounds),
            Err(view) => view.event(event, offset, bounds),
        }
    }

    fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
        match self {
            Ok(view) => view.draw(bounds, onto),
            Err(view) => view.draw(bounds, onto),
        }
    }
}

/// [`View`] events.
#[allow(missing_docs)]
#[derive(Copy, Clone, From, TryInto)]
pub enum Event {
    Gesture(Gesture),
    Resize { width: u32, height: u32 },
    Redraw,
}

/// touches… buttons…
#[derive(Copy, Clone)]
pub enum Gesture {
    Began { n: u8 },
    Moved { n: u8 },
    Ended { n: u8 },
}
