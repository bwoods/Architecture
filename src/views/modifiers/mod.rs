pub mod fixed;
pub mod offset;
pub mod padding;
