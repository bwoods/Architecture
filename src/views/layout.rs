#![allow(unused_imports)]
use crate::views::{Bounds, Event, Output, Point, Size, View};

#[doc(hidden)]
pub trait Layout {
    fn size(size: Size, next: Size) -> Size;
    fn bounds(bounds: Bounds, previous: Size) -> Bounds;
}

struct Vertical;

impl Layout for Vertical {
    fn size(size: Size, next: Size) -> Size {
        Size::new(f32::max(size.width, next.width), size.height + next.height)
    }

    fn bounds(mut bounds: Bounds, previous: Size) -> Bounds {
        bounds.min.y += previous.height;
        bounds
    }
}

struct Horizontal;

impl Layout for Horizontal {
    fn size(size: Size, next: Size) -> Size {
        Size::new(size.width + next.width, f32::max(size.height, next.height))
    }

    fn bounds(mut bounds: Bounds, previous: Size) -> Bounds {
        bounds.min.x += previous.width;
        bounds
    }
}

macro_rules! tuple_impl {
    ( $($val:ident)+ ) => {
        #[doc(hidden)]
        #[allow(non_snake_case)]
        #[allow(unused_variables)]
        impl<$($val: View),+> View for ( $($val,)+ ) {
            fn size(&self) -> Size {
                let ( $(ref $val,)+ ) = self;
                let size = Size::zero();

                $(
                    let size = Vertical::size(size, $val.size());
                )+

                size
            }

            fn event(&self, event: Event, offset: Point, bounds: Bounds) {
                let ( $(ref $val,)+ ) = self;

                $(
                    $val.event(event, offset, bounds);
                    let bounds = Vertical::bounds(bounds, $val.size());
                )+
            }

            fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
                let ( $(ref $val,)+ ) = self;

                $(
                    $val.draw(bounds, onto);
                    let bounds = Vertical::bounds(bounds, $val.size());
                )+
            }
        }
    };
}

tuple_impl! { A }
tuple_impl! { A B }
tuple_impl! { A B C }
tuple_impl! { A B C D }
tuple_impl! { A B C D E }
tuple_impl! { A B C D E F }
tuple_impl! { A B C D E F G }
tuple_impl! { A B C D E F G H }
tuple_impl! { A B C D E F G H I }
tuple_impl! { A B C D E F G H I J }
tuple_impl! { A B C D E F G H I J K }
tuple_impl! { A B C D E F G H I J K L }
tuple_impl! { A B C D E F G H I J K L M }
tuple_impl! { A B C D E F G H I J K L M N }
tuple_impl! { A B C D E F G H I J K L M N O }
tuple_impl! { A B C D E F G H I J K L M N O P }
tuple_impl! { A B C D E F G H I J K L M N O P Q }
tuple_impl! { A B C D E F G H I J K L M N O P Q R }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W X }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W X Y }
// up to 25 views are supported
