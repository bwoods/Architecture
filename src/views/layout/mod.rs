#![allow(unused_imports)]
use crate::views::{Bounds, Event, Output, Point, Size, View}; // some of these are used in the macro

pub use spacing::Spacer;

mod spacing;

#[doc(hidden)]
pub trait Layout {
    fn size(size: Size, next: Size) -> Size;
    fn bounds(bounds: Bounds, previous: Size) -> Bounds;
    fn space(n: u32, required: Size, available: Size) -> Size;
}

struct Vertical;

impl Layout for Vertical {
    fn size(size: Size, next: Size) -> Size {
        Size::new(f32::max(size.width, next.width), size.height + next.height)
    }

    fn bounds(mut bounds: Bounds, previous: Size) -> Bounds {
        bounds.min.y += previous.height;
        bounds.min.y = f32::min(bounds.min.y, bounds.max.y);
        bounds
    }

    fn space(n: u32, required: Size, mut available: Size) -> Size {
        let surplus = available.height - required.height;

        if surplus <= 0.0 {
            Size::zero()
        } else {
            available.height = surplus / n as f32;
            available.width = 0.0;
            available
        }
    }
}

struct Horizontal;

impl Layout for Horizontal {
    fn size(size: Size, next: Size) -> Size {
        Size::new(size.width + next.width, f32::max(size.height, next.height))
    }

    fn bounds(mut bounds: Bounds, previous: Size) -> Bounds {
        bounds.min.x += previous.width;
        bounds.min.x = f32::min(bounds.min.x, bounds.max.x);
        bounds
    }

    fn space(n: u32, required: Size, mut available: Size) -> Size {
        let surplus = available.width - required.width;

        if surplus <= 0.0 {
            Size::zero()
        } else {
            available.width = surplus / n as f32;
            available.height = 0.0;
            available
        }
    }
}

///
#[doc(hidden)]
struct Row<T>(T);

macro_rules! tuple_impl {
    ( $($val:ident)+ ) => {
        #[doc(hidden)]
        #[allow(non_snake_case)]
        #[allow(unused_variables)]
        impl<$($val: View),+> View for ( $($val,)+ ) {
            #[inline]
            fn size(&self) -> Size {
                let ( $(ref $val,)+ ) = self;
                let size = Size::zero();

                $(
                    let size = Vertical::size(size, $val.size());
                )+

                size
            }

            #[inline]
            fn event(&self, event: Event, offset: Point, bounds: Bounds) {
                self.update_layout(self.size(), bounds);

                let ( $(ref $val,)+ ) = self;
                $(
                    let size = $val.size();
                    $val.event(event, offset, bounds);
                    let bounds = Vertical::bounds(bounds, size);
                )+
            }

            #[inline]
            fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
                self.update_layout(self.size(), bounds);

                let ( $(ref $val,)+ ) = self;
                $(
                    let size = $val.size();
                    $val.draw(bounds, onto);
                    let bounds = Vertical::bounds(bounds, size);
                )+
            }

            fn update_layout(&self, size: Size, bounds: Bounds) {
                let ( $(ref $val,)+ ) = self;
                let mut size = Size::zero();
                let mut n = 0;

                $(
                    size = Vertical::size(size, $val.size());
                    n += $val.needs_layout() as u32;
                )+

                if n != 0 {
                    let space = Vertical::space(n, size, bounds.size());
                    $(
                        if $val.needs_layout() {
                            $val.update_layout(space, bounds);
                        }
                    )+;
                }
            }

            #[inline(always)]
            fn across(self) -> impl View {
                Row(self)
            }
        }

        #[doc(hidden)]
        #[allow(non_snake_case)]
        #[allow(unused_variables)]
        impl<$($val: View),+> View for Row<( $($val,)+ )> {
            fn size(&self) -> Size {
                let ( $(ref $val,)+ ) = self.0;
                let size = Size::zero();

                $(
                    let size = Horizontal::size(size, $val.size());
                )+

                size
            }

            fn event(&self, event: Event, offset: Point, bounds: Bounds) {
                let ( $(ref $val,)+ ) = self.0;

                $(
                    let size = $val.size();
                    $val.event(event, offset, bounds);
                    let bounds = Horizontal::bounds(bounds, size);
                )+
            }

            fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
                let ( $(ref $val,)+ ) = self.0;

                $(
                    let size = $val.size();
                    $val.draw(bounds, onto);
                    let bounds = Horizontal::bounds(bounds, size);
                )+
            }

            #[inline(always)]
            fn needs_layout(&self) -> bool {
                self.0.needs_layout()
            }

            #[inline(always)]
            fn update_layout(&self, size: Size, bounds: Bounds) {
                let ( $(ref $val,)+ ) = self.0;
                let size = Size::zero();
                let n = 0;

                $(
                    let size = Horizontal::size(size, $val.size());
                    let n = n + $val.needs_layout() as u32;
                )+

                if n != 0 {
                    let space = Horizontal::space(n, size, bounds.size());
                    $(
                        if $val.needs_layout() {
                            $val.update_layout(space, bounds);
                        }
                    )+;
                }
            }
        }
    };
}

tuple_impl! { A }
tuple_impl! { A B }
tuple_impl! { A B C }
tuple_impl! { A B C D }
tuple_impl! { A B C D E }
tuple_impl! { A B C D E F }
tuple_impl! { A B C D E F G }
tuple_impl! { A B C D E F G H }
tuple_impl! { A B C D E F G H I }
tuple_impl! { A B C D E F G H I J }
tuple_impl! { A B C D E F G H I J K }
tuple_impl! { A B C D E F G H I J K L }
tuple_impl! { A B C D E F G H I J K L M }
tuple_impl! { A B C D E F G H I J K L M N }
tuple_impl! { A B C D E F G H I J K L M N O }
tuple_impl! { A B C D E F G H I J K L M N O P }
tuple_impl! { A B C D E F G H I J K L M N O P Q }
tuple_impl! { A B C D E F G H I J K L M N O P Q R }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W X }
tuple_impl! { A B C D E F G H I J K L M N O P Q R S T U V W X Y }
// up to 25 views are supported
