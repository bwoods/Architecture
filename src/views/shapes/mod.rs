use crate::dependencies::Dependency;
use crate::views::{Bounds, Output, Size, Transform, View};

mod rounded;

pub trait Path: Sized {
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output);

    fn fill(self) -> Shape<Self> {
        self.fixed(f32::INFINITY, f32::INFINITY)
    }

    fn fixed(self, width: f32, height: f32) -> Shape<Self> {
        Shape {
            size: Size::new(width, height),
            path: self,
        }
    }
}

/// [Least-squares approximation of the circle using cubic Bézier curves][site]
///
/// > David Ellsworth found the optimal value of c:  
/// >
/// > c ≈ 0.5519703814011128603134107
///
/// [site]: https://spencermortensen.com/articles/least-squares-bezier-circle/
pub(crate) const K: f32 = 0.4480296; // 1 - 0.5519703814011128603134107 rounded to f32

pub struct Rectangle {
    pub rgba: [u8; 4],
}

impl Path for Rectangle {
    #[inline(always)]
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output) {
        rounded::rectangle(x, y, w, h, 0.0, 0.0, 0.0, self.rgba, transform, onto);
    }
}

impl Rectangle {
    pub fn rounded(self, rx: f32, ry: f32) -> RoundedRectangle {
        RoundedRectangle {
            rgba: self.rgba,
            rx,
            ry,
        }
    }
}

pub struct RoundedRectangle {
    rgba: [u8; 4],
    rx: f32,
    ry: f32,
}

impl Path for RoundedRectangle {
    #[inline(always)]
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output) {
        rounded::rectangle(x, y, w, h, self.rx, self.ry, K, self.rgba, transform, onto);
    }
}

impl RoundedRectangle {
    pub fn continuous(self) -> ContinuousRoundedRectangle {
        ContinuousRoundedRectangle {
            rgba: self.rgba,
            rx: self.rx,
            ry: self.ry,
        }
    }
}

pub struct ContinuousRoundedRectangle {
    rgba: [u8; 4],
    rx: f32,
    ry: f32,
}

impl Path for ContinuousRoundedRectangle {
    #[inline(always)]
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output) {
        // continuous corners are much smaller than circular ones; scale them up a bit
        let c = std::f32::consts::E;
        let rx = (self.rx * c).min(w / 2.0);
        let ry = (self.ry * c).min(h / 2.0);
        rounded::rectangle(x, y, w, h, rx, ry, 0.0, self.rgba, transform, onto);
    }
}

pub struct Ellipse {
    pub rgba: [u8; 4],
}

impl Path for Ellipse {
    #[inline(always)]
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output) {
        let rx = w / 2.0;
        let ry = h / 2.0;
        rounded::rectangle(x, y, w, h, rx, ry, K, self.rgba, transform, onto);
    }
}

pub struct Circle {
    pub rgba: [u8; 4],
}

impl Path for Circle {
    #[inline(always)]
    fn draw(&self, x: f32, y: f32, w: f32, h: f32, transform: &Transform, onto: &mut impl Output) {
        let r = f32::min(w, h) / 2.0;
        rounded::rectangle(x, y, w, h, r, r, K, self.rgba, transform, onto);
    }
}

#[doc(hidden)]
pub struct Shape<T> {
    size: Size,
    path: T,
}

impl<T: Path> View for Shape<T> {
    #[inline(always)]
    fn size(&self) -> Size {
        self.size
    }

    #[inline]
    fn draw(&self, bounds: Bounds, onto: &mut impl Output) {
        let size = match (self.size.width.is_finite(), self.size.height.is_finite()) {
            (true, true) => self.size,
            (false, false) => bounds.size(),
            (true, false) => Size::new(self.size.width, bounds.height()),
            (false, true) => Size::new(bounds.width(), self.size.height),
        };

        self.path.draw(
            bounds.min.x,
            bounds.min.y,
            size.width,
            size.height,
            &Dependency::<Transform>::new().unwrap_or_default(),
            onto,
        );
    }

    #[inline(always)]
    #[allow(refining_impl_trait)]
    fn fixed(mut self, width: f32, height: f32) -> Self {
        self.size = Size::new(width, height);
        self
    }

    #[inline(always)]
    #[allow(refining_impl_trait)]
    fn width(mut self, width: f32) -> Self {
        self.size.width = width;
        self
    }

    #[inline(always)]
    #[allow(refining_impl_trait)]
    fn height(mut self, height: f32) -> Self {
        self.size.height = height;
        self
    }
}
