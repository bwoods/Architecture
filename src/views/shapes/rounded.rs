use crate::views::{Output, Transform};

#[inline(never)]
#[allow(clippy::too_many_arguments)]
pub(crate) fn rectangle(
    x: f32,
    y: f32,
    w: f32,
    h: f32,
    rx: f32,
    ry: f32,
    k: f32,
    rgba: [u8; 4],
    transform: &Transform,
    output: &mut impl Output,
) {
    let p0 = (x, y + ry);
    let c0 = (x, y + ry * k);
    let c1 = (x + rx * k, y);
    let p1 = (x + rx, y);

    let p2 = (x + w - rx, y);
    let c2 = (x + w - rx * k, y);
    let c3 = (x + w, y + ry * k);
    let p3 = (x + w, y + ry);

    let p4 = (x + w, y + h - ry);
    let c4 = (x + w, y + h - ry * k);
    let c5 = (x + w - rx * k, y + h);
    let p5 = (x + w - rx, y + h);

    let p6 = (x + rx, y + h);
    let c6 = (x + rx * k, y + h);
    let c7 = (x, y + h - ry * k);
    let p7 = (x, y + h - ry);

    output.begin(p0.0, p0.1, rgba, transform);
    output.cubic_bezier_to(c0.0, c0.1, c1.0, c1.1, p1.0, p1.1);
    output.line_to(p2.0, p2.1);
    output.cubic_bezier_to(c2.0, c2.1, c3.0, c3.1, p3.0, p3.1);
    output.line_to(p4.0, p4.1);
    output.cubic_bezier_to(c4.0, c4.1, c5.0, c5.1, p5.0, p5.1);
    output.line_to(p6.0, p6.1);
    output.cubic_bezier_to(c6.0, c6.1, c7.0, c7.1, p7.0, p7.1);
    output.close();
}

#[test]
fn snapshot_testing() {
    use super::Path;
    use insta::assert_snapshot;

    let black = [0, 0, 0, 0xff];
    let transform = Default::default();

    let mut output = crate::views::svg::Output::new(256.0, 256.0);
    let circle = super::Circle { rgba: black };
    circle.draw(16.0, 16.0, 224.0, 224.0, &transform, &mut output);
    assert_snapshot!("circle", output.into_inner());

    let mut output = crate::views::svg::Output::new(256.0, 128.0);
    let ellipse = super::Ellipse { rgba: black };
    ellipse.draw(16.0, 16.0, 224.0, 112.0, &transform, &mut output);
    assert_snapshot!("ellipse", output.into_inner());

    let mut output = crate::views::svg::Output::new(256.0, 128.0);
    let rectangle = super::Rectangle { rgba: black };
    rectangle.draw(16.0, 16.0, 224.0, 112.0, &transform, &mut output);
    assert_snapshot!("rectangle", output.into_inner());

    let mut output = crate::views::svg::Output::new(256.0, 128.0);
    let rounded = rectangle.rounded(16.0, 16.0);
    rounded.draw(16.0, 16.0, 224.0, 112.0, &transform, &mut output);
    assert_snapshot!("rounded", output.into_inner());

    let mut output = crate::views::svg::Output::new(256.0, 128.0);
    let continuous = rounded.continuous();
    continuous.draw(16.0, 16.0, 224.0, 112.0, &transform, &mut output);
    assert_snapshot!("continuous", output.into_inner());
}
