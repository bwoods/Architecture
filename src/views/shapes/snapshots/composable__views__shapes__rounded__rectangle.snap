---
source: src/views/shapes/rounded.rs
expression: output.into_inner()
---
<svg height="128" viewBox="0 0 256 128" width="256" xmlns="http://www.w3.org/2000/svg">
<path d="M16,16 C16,16,16,16,16,16 L240,16 C240,16,240,16,240,16 L240,128 C240,128,240,128,240,128 L16,128 C16,128,16,128,16,128 z" fill="#000000ff" transform="matrix(1 0 0 1 0 0)"/>
</svg>
